x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStopDatafeedAction.java

```diff
@@ -211,7 +218,64 @@ private void doExecute(
+                            if (datafeedTask != null
+                                && PersistentTasksClusterService.needsReassignment(datafeedTask.getAssignment(), nodes) == false) {
+                                startedDatafeedsJobs.add(((StartDatafeedAction.DatafeedParams) datafeedTask.getParams()).getJobId());
+                            }
+                        }
+                        if (request.closeJob() && startedDatafeedsJobs.isEmpty() == false) {
+                            // If the "close_job" parameter was set to "true" on the stop datafeed request we attempt to first close the
+                            // jobs associated with the datafeeds. This will in turn attempt to stop the jobs' datafeeds (this time with the
+                            // "close_job" flag set to false, to avoid recursion)
+                            ActionListener<List<Tuple<String, AcknowledgedResponse>>> closeJobActionListener = listener
+                                .delegateFailureAndWrap((delegate, jobsResponses) -> {
+                                    List<String> jobIds = jobsResponses.stream()
+                                        .filter(t -> t.v2().isAcknowledged() == false)
+                                        .map(Tuple::v1)
+                                        .collect(toList());
+                                    if (jobIds.isEmpty()) {
+                                        logger.debug("Successfully closed jobs (and associated datafeeds)");
+                                    } else {
+                                        logger.warn("Failed to close jobs (and associated datafeeds): {}", jobIds);
+                                    }
+                                    delegate.onResponse(new StopDatafeedAction.Response(true));
+                                });
+
+                            TypedChainTaskExecutor<Tuple<String, AcknowledgedResponse>> chainTaskExecutor = new TypedChainTaskExecutor<>(
+                                EsExecutors.DIRECT_EXECUTOR_SERVICE,
+                                Predicates.always(),
+                                Predicates.always()
+                            );
+                            for (String jobId : startedDatafeedsJobs) {
+                                chainTaskExecutor.add(
+                                    al -> executeAsyncWithOrigin(
+                                        client,
+                                        ML_ORIGIN,
+                                        CloseJobAction.INSTANCE,
+                                        new CloseJobAction.Request(jobId),
+                                        listener.delegateFailureAndWrap(
+                                            (l, response) -> l.onResponse(new StopDatafeedAction.Response(response.isClosed()))
+                                        )
```

Comment:
```suggestion
                  al.delegateFailureAndWrap((l, response) -> {
                l.onResponse(Tuple.tuple(jobId, response));
            }, (l, e) -> {
                Throwable unwrapped = ExceptionsHelper.unwrapCause(e);
                if (unwrapped instanceof ResourceNotFoundException) {
               logger.debug(
                        () -> format(
                            "[%s] Job was already closed or does not exist when attempting to close via stop datafeed",
                            jobId
                        ),
                        e
                    );
                    l.onResponse(Tuple.tuple(jobId, new CloseJobAction.Response(true)));
                } else {
                   l.onFailure(e);
                }
            })
```

If, meanwhile, the job was already closed by other means, treat it as a success to avoid race conditions.